File: app/src/main/java/me/evrooij/groceries/network/ClientInterface.java
package me.evrooij.groceries.network;

import me.evrooij.groceries.models.Account;
import me.evrooij.groceries.models.Feedback;
import me.evrooij.groceries.models.GroceryList;
import me.evrooij.groceries.models.Product;
import retrofit2.Call;
import retrofit2.http.*;

import java.util.List;
import java.util.Map;

/**
 * Created by eddy
 * Date: 11-12-16.
 */

public interface ClientInterface {

    // Login
    @GET("/users/login")
    Call<Account> getAccountByLogin(@Query("username") String username, @Query("password") String password);

    @POST("/users/register")
    Call<Account> registerAccount(@Body Map<String, String> accountMap);


    // Friends
    @GET("/accounts/{ownId}/friends/find")
    Call<List<Account>> findFriends(@Path("ownId") int ownId, @Query("query") String searchQuery);

    @POST("/accounts/{ownId}/friends/add")
    Call<ResponseMessage> addFriend(@Path("ownId") int ownId, @Body Account newFriend);

    @GET("/accounts/{ownId}/friends")
    Call<List<Account>> getFriends(@Path("ownId") int ownId);


    // Lists
    @GET("/user/{id}/lists")
    Call<List<GroceryList>> getLists(@Path("id") int accountId);

    @GET("/lists/{id}")
    Call<GroceryList> getList(@Path("id") int listId);

    @POST("/lists/new")
    Call<GroceryList> newList(@Body GroceryList list);


    // Products
    @POST("/list/{id}/products/new")
    Call<Product> newProduct(@Path("id") int listId, @Body Product newProduct);

    @DELETE("/lists/{listId}/products/{productId}")
    Call<ResponseMessage> deleteProduct(@Path("listId") int listId, @Path("productId") int productId);

    @PUT("/lists/{listId}/products/{productId}/edit")
    Call<ResponseMessage> editProduct(@Path("listId") int listId, @Path("productId") int productId, @Body Product editedProduct);

    @GET("/{id}/myproducts")
    Call<List<Product>> getMyProducts(@Path("id") int accountId);

    @POST("/{id}/myproducts")
    Call<Product> addMyProduct(@Path("id") int accountId, @Body Product product);


    // Feedback
    @POST("/feedback/new")
    Call<ResponseMessage> reportFeedback(@Body Feedback feedback);

}

File: app/src/main/java/me/evrooij/groceries/network/ClientInterface.kt
package me.evrooij.groceries.network

import me.evrooij.groceries.models.Account
import me.evrooij.groceries.models.Feedback
import me.evrooij.groceries.models.GroceryList
import me.evrooij.groceries.models.Product
import retrofit2.Call
import retrofit2.http.*

/**
 * Created by eddy
 * Date: 11-12-16.
 */
interface ClientInterface {
    // Login
    @GET("/users/login")
    fun getAccountByLogin(@Query("username") username: String, @Query("password") password: String): Call<Account>

    @POST("/users/register")
    fun registerAccount(@Body accountMap: Map<String, String>): Call<Account>


    // Friends
    @GET("/accounts/{ownId}/friends/find")
    fun findFriends(@Path("ownId") ownId: Int, @Query("query") searchQuery: String): Call<List<Account>>

    @POST("/accounts/{ownId}/friends/add")
    fun addFriend(@Path("ownId") ownId: Int, @Body newFriend: Account): Call<ResponseMessage>

    @GET("/accounts/{ownId}/friends")
    fun getFriends(@Path("ownId") ownId: Int): Call<List<Account>>


    // Lists
    @GET("/user/{id}/lists")
    fun getLists(@Path("id") accountId: Int): Call<List<GroceryList>>

    @GET("/lists/{id}")
    fun getList(@Path("id") listId: Int): Call<GroceryList>

    @POST("/lists/new")
    fun newList(@Body list: GroceryList): Call<GroceryList>


    // Products
    @POST("/list/{id}/products/new")
    fun newProduct(@Path("id") listId: Int, @Body newProduct: Product): Call<Product>

    @DELETE("/lists/{listId}/products/{productId}")
    fun deleteProduct(@Path("listId") listId: Int, @Path("productId") productId: Int): Call<ResponseMessage>

    @PUT("/lists/{listId}/products/{productId}/edit")
    fun editProduct(@Path("listId") listId: Int, @Path("productId") productId: Int, @Body editedProduct: Product): Call<ResponseMessage>

    @GET("/{id}/myproducts")
    fun getMyProducts(@Path("id") accountId: Int): Call<List<Product>>

    @POST("/{id}/myproducts")
    fun addMyProduct(@Path("id") accountId: Int, @Body product: Product): Call<Product>


    // Feedback
    @POST("/feedback/new")
    fun reportFeedback(@Body feedback: Feedback): Call<ResponseMessage>

}
